add task (#1)
mate-academy/py-token-authentication: README.md

# User Permissions API

- Read [the guideline](https://github.com/mate-academy/py-task-guideline/blob/main/README.md) before start
- Download [ModHeader](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj?hl=en)

### In this task you will add the functionality of user permissions

1. Create serializers and views to support the following endpoints:
   * `POST api/register/` - You can create here a user
   * `POST api/login/` - You can get a token, if you write the correct data
   * `POST api/me/` - Information about user and possibility to update information about user

Example:
```python
HTTP 200 OK
Allow: GET, PUT, PATCH, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "id": 1,
    "username": "admin1",
    "email": "",
    "is_staff": true
}
```

> 2. Create `cinema/permissions.py` with next API access:

Reviewer: By default, all endpoints in API must have the following action limitations depending on the user role:
```suggestion
2. By default, all endpoints in API must have the following action limitations depending on the user role:
```

Author: done

 * Anon: None.
 * IsAuthenticated: list, retrieve.

>  * IsAdmin: create, update, partial_update, destroy.

Reviewer: Implement such custom permission class `IsAdminOrIfAuthenticatedReadOnly`

Author: dome

3. Add `authentication_classes` & `permission_classes` for all view classes.

> 4. Use mixins for `views.py` with next access:

Reviewer: ```suggestion
4. Make **only** such actions available for views:
```

Author: done

   * `GenreViewSet` - list and create
   * `CinemaHallViewSet` - list and create
   * `ActorViewSet` - list and create 
   * `MovieViewSet` - list, create and retrieve

>    * `MovieSessionViewSet` - without changes

Reviewer: ```suggestion
   * `MovieSessionViewSet` - list, retrieve, create, update, partial_update, delete
```

Author: done

   * `OrderViewSet` - list and create

> 5. `OrderViewSet` - implements `get_permissions`. This function allows to create an order.

Reviewer: This is an incorrect description.
Do not give them hint about `get_permissions`.
Say, that for Orders - we want give all authenticated users ability to create an order (for that user of course)

Author: done

# Token Authentication API

- Read [the guideline](https://github.com/mate-academy/py-task-guideline/blob/main/README.md) before start

> - Download [ModHeader](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj?hl=en)

Reviewer: Link to ModHeader also should be placed in theory for this topic

### In this task you will add the functionality of token authentication

> At this part of the task, we will do authorization by using tokens. The functionality of regular users will be limited so that they cannot add, delete or update other data on the sites, besides their orders.  Moreover, only authenticated users will be able to create an order.  Deletion will be prohibited even for the administrator, if only through the admin panel. That's because of  when we're deleting, for example, a genre, the other relationships from other tables won't deleted

Reviewer: ```suggestion
At this part of the task, we will do authorization by using tokens. The functionality of regular users will be limited so that they cannot add, delete or update other data on the site, besides their orders.  Moreover, only authenticated users will be able to create an order.  Deletion will be prohibited even for the administrator, if only through the admin panel. That's because of  when we're deleting, for example, a genre, the other relationships from other tables won't be deleted
```

1. Create serializers and views to support the following endpoints:

>    * `POST api/register/` - You can create here a user

Reviewer: ```suggestion
   * `POST api/user/register/` - You can create here a user (password length must be >= 5 symbols)
```

>    * `POST api/login/` - You can get a token, if you write the correct data

Reviewer: ```suggestion
   * `POST api/user/login/` - You can get a token, if you write the correct data
```

>    * `PATCH api/me/` - Information about user and possibility to update information about user

Reviewer: ```suggestion
   * `GET/PUT/PATCH api/user/me/` - Information about user and possibility to update information about user
```

Example:
```python
HTTP 200 OK
Allow: GET, PUT, PATCH, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "id": 1,
    "username": "admin1",
    "email": "",
    "is_staff": true
}
```

> 2. By default, all endpoints in API must have the following action limitations depending on the user role:

Reviewer: ```suggestion
2. By default, all API endpoints (inside cinema app) must have the following action limitations depending on the user role:
```

 * Implement such custom permission class `IsAdminOrIfAuthenticatedReadOnly`.

>  * Password should have `min_length = 5`

Reviewer: ```suggestion
```

> 3. Add `authentication_classes` & `permission_classes` for all view classes.

Reviewer: ```suggestion
```

> 4. Make **only** such actions available for views:

Reviewer: ```suggestion
3. Make **only** such actions available for views:
```

   * `GenreViewSet` - list and create
   * `CinemaHallViewSet` - list and create
   * `ActorViewSet` - list and create 
   * `MovieViewSet` - list, create and retrieve
   * `MovieSessionViewSet` - list, retrieve, create, update, partial_update, delete
   * `OrderViewSet` - list and create

> 5. `OrderViewSet` - We should give the ability for authenticated users to create order

Reviewer: ```suggestion
4. `OrderViewSet` - We should give the ability for authenticated users to create order
```

> `Note` all tests should are passed. `user/tests` & `cinema/tests`

Reviewer: ```suggestion
`Note` all tests should pass. `user/tests` & `cinema/tests`
```
